### PR TITLE
implement ACK_interval_respond

### PR DESCRIPTION
**Pull request by sbw111**
implement ACK_interval_respond & error control

more test needed.....

### CODE REVIEW

File: netsec_fall2017/lab2/src/lab2_protocol/PEEPClientProtocol.py [hunk 1]
```diff
@@ -86,24 +85,32 @@ def connection_lost(self, exc=None):
 	def __data_packet_handler(self,packet):
 		if self.state != "Transmission_State_2":
 			if self.logging:
-				print("PEEP Client Side: Error: State Error! Expecting Transmission_State_2 but getting %s" % self.state)
+				print("PEEP Server Side: Error: State Error! Expecting Transmission_State_2 but getting %s" % self.state)
```
Comment (Zifan94) [inline netsec_fall2017/lab2/src/lab2_protocol/PEEPClientProtocol.py:88]: "Client Side" I think

File: netsec_fall2017/lab2/src/lab2_protocol/PEEPClientProtocol.py [hunk 2]
```diff
@@ -86,24 +85,32 @@ def connection_lost(self, exc=None):
...
-			outBoundPacket = Util.create_outbound_packet(2, None, packet.SequenceNumber + len(packet.Data))
-			packetBytes = outBoundPacket.__serialize__()
-			if self.logging:
-				print("PEEP Client Side: ACK back <=")
-			#####################################################
-
-			self.transport.write(packetBytes)
-			self.higherProtocol().data_received(packet.Data)
-			
+				print("PEEP Server Side: Data Chunck reveived: Seq = %d, Checksum = (%d)" % (packet.SequenceNumber, packet.Checksum))
```
Comment (Zifan94) [inline netsec_fall2017/lab2/src/lab2_protocol/PEEPClientProtocol.py:92]: "Client Side"

File: netsec_fall2017/lab2/src/lab2_transport/PEEPTransport.py [hunk 3]
```diff
@@ -23,25 +31,56 @@ def write(self, data):
...
+	def retransmission_checker(self,seq):
+		if seq in self.RetransmissionPacketList:
+			if self.logging:
+				print("PEEP Transport: Packets ack = [%s] not received after TIMEOUT, Retransmission...." %seq)
+			self.lowerTransport().write(self.RetransmissionPacketList[seq].__serialize__())
+			asyncio.get_event_loop().call_later(self.TIME_OUT_LIMIE, self.retransmission_checker, seq)
 
 	def ack_received(self,ack):
 		if self.logging:	print("PEEP Transport: ACK received, Ack = %d" % ack)
+		while (self.ackList[1] < ack):
```
Comment (Zifan94) [inline netsec_fall2017/lab2/src/lab2_transport/PEEPTransport.py:61]: you could change "<" into "<=" and delete the "if statement" below

File: netsec_fall2017/lab2/src/lab2_protocol/PEEPClientProtocol.py (hunk 2 above)
Comment (Zifan94) [inline netsec_fall2017/lab2/src/lab2_protocol/PEEPClientProtocol.py:92]: "Client Side"

File: netsec_fall2017/lab2/src/lab2_protocol/PEEPClientProtocol.py [hunk 4]
```diff
@@ -86,24 +85,32 @@ def connection_lost(self, exc=None):
...
+
+			# if packet.Acknowledgement != None:
+			# 	self.__ack_handler(packet.Acknowledgement)
+			print(self.seq_except)
```
Comment (Zifan94) [inline netsec_fall2017/lab2/src/lab2_protocol/PEEPClientProtocol.py:96]: seq_expected I guess?

**Review by Zifan94 — COMMENTED** [review]
I added some comment about some small issue need to fixed.
Other parts looks good to me!
I ran it on my local machine and tested with a smaller ACK_TIME_INTERVAL which looks great! (The smaller ack time interval is, the more ack we can see during the communication!)

Good job!

File: netsec_fall2017/lab2/src/lab2_transport/PEEPTransport.py (hunk 3 above)
Comment (sbw111) [reply netsec_fall2017/lab2/src/lab2_transport/PEEPTransport.py:61]: I can't do that, beacuse it will cause a bug when delete the last element.

**Pull request closed without merging**